README.md
# opsline-rails-app Cookbook
This cookbook has been created to install, configure, and manage rails applications.
Applications are installed from artifacts stored in S3. All configuration parameters
are stored in data bag items.


# Requirements
#### cookbooks
- `artifact` - Artifact Deploy LWRP
- `ruby_build`
- `rbenv`
- `rvm`
- `logrotate`


# Attributes
* `node['opsline-rails-app']['databag']` - data bag name to store app details
* `node['opsline-rails-app']['apps']` - list of apps to deploy
* `node['opsline-rails-app']['encrypted_databag']` - true to use encrypted data bags
* `node['opsline-rails-app']['s3_bucket']` - name of the S3 bucket that holds artifacts
* `node['opsline-rails-app']['owner']` - unix user used to own and run applications
* `node['opsline-rails-app']['apps_root']` - root directory for applications
* `node['opsline-rails-app']['unicorn']['timeout']` - unicorn timeout
* `node['opsline-rails-app']['unicorn']['worker_processes']` - number of unicorn workers
* `node['opsline-rails-app']['ruby']['provider']` - rvm or rbenv
* `node['opsline-rails-app']['ruby']['versions']` - ruby versions
* `node['opsline-rails-app']['ruby']['gems']` - ruby gems


#Usage
#### opsline-rails-app::default
Prepares for deployment:
* installs required packages
* configured ruby if configured

#### opsline-rails-app::deploy
Deployed rails applications.

Rails applications are defined as data bag items in a data bag defined in attributes.

Example of app definition:
```json
{
  "id": "testapp",
  "name": "testapp",
  "artifact_name": "testapp",
  "type": "rails",
  "container": "unicorn",
  "container_parameters": {
    "default": {
      "worker_processes": "4",
      "frontend": "nginx"
    }
  },
  "version": {
    "production": "1",
    "default": "2"
  },
  "environment": {
    "production": {
      "REDIS_URL": "redis://redis-prod.example.com:6379",
      "PGBACKUPS_URL": "https://user:pass@postgresql-prod.example.com/schema"
    },
    "default": {
      "REDIS_URL": "redis://redis-test.example.com:6379",
      "PGBACKUPS_URL": "https://user:pass@postgresql-test.example.com/schema"
    }
  },
  "packages": [
    "nodejs",
    "imagemagick",
    "openjdk-7-jre-headless"
  ]
}
```

Define a list of rails apps to deploy in attributes.
```ruby
name "rails-testapp"
description "rails test app"
run_list(
  "role[base]",
  "role[ruby]",
  "recipe[opsline-rails-app::default]",
  "role[nginx]",
  "recipe[opsline-rails-app::deploy]"
)
default_attributes(
  opsline-rails-app" => {
    "owner" => "rails",
    "s3_bucket" => "example.artifacts",
    "apps" => ["testapp"],
    "ruby" => {
      "provider" => "rvm"
    }
  }
)
```

Valid containers are:
* unicorn
* rack
* passenger
* worker


License and Authors
-------------------
* Author:: Radek Wierzbicki

```text
Copyright 2014, OpsLine, LLC.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
```


CHANGELOG.md
# CHANGELOG for opsline-mysql

## 0.6.4
* updating README.md
* applying Apache License

## 0.6.3
* adding WORKER_NAME variable for workers

## 0.6.2
* fixing inheritance

## 0.6.1
* fixing env_dir definition

## 0.6.0
* adding app data bag inheritance
* removed environment variables will be deleted

## 0.5.2
* nginx setup rewrite

## 0.5.1
* Adding logrotate

## 0.5.0
* Adding worker container
* Ability to install multiple versions of ruby via rbenv
* Support for rvm
* Using encrypted data bags

## 0.1.0:
* Initial release


recipes/default.rb
#
# Cookbook Name:: opsline-rails-app
# Recipe:: default
#
# Author:: Radek Wierzbicki
#
# Copyright 2014, OpsLine, LLC.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

class Chef::Recipe
  include Opsline::RailsApp::Helpers
end

# install required packages
package 'daemontools'
package 'inotify-tools'

# create owner if does not exist
user node['opsline-rails-app']['owner'] do
  action :create
  comment "Rails User"
  supports :manage_home => true
  home "/home/#{node['opsline-rails-app']['owner']}"
  shell '/bin/bash'
  not_if "id #{node['opsline-rails-app']['owner']} >/dev/null 2>&1"
end

if node['opsline-rails-app']['ruby']['provider'] == 'rvm'
  unless node['opsline-rails-app']['ruby']['versions'].empty?
    include_recipe 'ruby_build'

    node.default['rvm']['user_installs'] = [
      {'user' => node['opsline-rails-app']['owner']}
    ]
    include_recipe 'rvm::user_install'

    node['opsline-rails-app']['ruby']['versions'].each do |ruby_version|
      rvm_ruby "ruby #{ruby_version}" do
        ruby_string ruby_version
        user node['opsline-rails-app']['owner']
        action :install
      end
      node['opsline-rails-app']['ruby']['gems'].each do |gem_name, gem_info|
        rvm_gem "#{gem_name} for ruby #{ruby_version}" do
          package_name gem_name
          version gem_info['version']
          ruby_string ruby_version
          user node['opsline-rails-app']['owner']
          action gem_info['ruby_versions'].include?(ruby_version) ? :install : :remove
        end
      end
    end
  end

elsif node['opsline-rails-app']['ruby']['provider'] == 'rbenv'
  unless node['opsline-rails-app']['ruby']['versions'].empty?
    include_recipe 'ruby_build'

    node.default['rbenv']['user_installs'] = [
      {'user' => node['opsline-rails-app']['owner']}
    ]
    include_recipe 'rbenv::user_install'

    node['opsline-rails-app']['ruby']['versions'].each do |ruby_version|
      rbenv_ruby "ruby #{ruby_version}" do
        definition ruby_version
        user node['opsline-rails-app']['owner']
        action :install
      end
      #rbenv_global node['opsline-rails-app']['ruby']['version'] do
      #  rbenv_version ruby_version
      #  user node['opsline-rails-app']['owner']
      #end
      node['opsline-rails-app']['ruby']['gems'].each do |gem_name, gem_info|
        rbenv_gem "#{gem_name} for ruby #{ruby_version}" do
          package_name gem_name
          version gem_info['version']
          rbenv_version ruby_version
          user node['opsline-rails-app']['owner']
          action gem_info['ruby_versions'].include?(ruby_version) ? :install : :remove
        end
      end
    end
  end

end

directory node['opsline-rails-app']['apps_root'] do
  action :create
  owner node['opsline-rails-app']['owner']
  group node['opsline-rails-app']['owner']
  mode 0755
end


definitions/frontend_nginx.rb
define :nginx_app_config, :app_name => nil, :app_data => {}, :template => 'nginx.conf.erb', :cookbook => 'opsline-rails-app', :action => :create do

  app_name = params[:app_name]
  app_data = params[:app_data]

  service 'nginx' do
    action :nothing
  end

  template "/etc/nginx/sites-available/#{app_name}" do
    source params[:template]
    cookbook params[:cookbook]
    owner 'root'
    group 'root'
    mode 0644
    action params[:action]
    notifies :restart, resources(:service => 'nginx'), :delayed
    variables({
      :app_name => app_name,
      :app_data => app_data,
      :server_name => "#{app_name}.#{node.domain}"
    })
  end
  link "#{app_name} nginx site" do
    target_file "/etc/nginx/sites-enabled/#{app_name}"
    to "/etc/nginx/sites-available/#{app_name}"
    action params[:action]
  end

end


recipes/deploy.rb
#
# Cookbook Name:: opsline-rails-app
# Recipe:: deploy
#
# Author:: Radek Wierzbicki
#
# Copyright 2014, OpsLine, LLC.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

class Chef::Recipe
  include Opsline::RailsApp::Helpers
end

# install all configured rails apps
node['opsline-rails-app']['apps'].each do |app_id|

  if node['opsline-rails-app']['encrypted_databag']
    app_data = Chef::EncryptedDataBagItem.load(node['opsline-rails-app']['databag'], app_id).to_hash
  else
    app_data = data_bag_item(node['opsline-rails-app']['databag'], app_id)
  end

  # read inherited data bag item
  if app_data.has_key?('inherits')
    if node['opsline-rails-app']['encrypted_databag']
      inherited_data = Chef::EncryptedDataBagItem.load(node['opsline-rails-app']['databag'], app_data['inherits']).to_hash
    else
      inherited_data = data_bag_item(node['opsline-rails-app']['databag'], app_data['inherits'])
    end
  else
    inherited_data = nil
  end

  # get app name
  if app_data.has_key?('name')
    app_name = app_data['name']
  else
    app_name = app_id
  end

  # initial environment variables hash
  app_data['app_env'] = {}
  app_data['app_env']['RACK_ENV'] = node.chef_environment
  app_data['app_env']['RAILS_ENV'] = node.chef_environment
  app_data['app_env']['HOME'] = "/home/#{node['opsline-rails-app']['owner']}"
  app_data['app_env']['PATH'] = "/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
  if node['opsline-rails-app']['ruby']['provider'] == 'rvm'
    app_data['app_env']['PATH'] = "/home/#{node['opsline-rails-app']['owner']}/.rvm/bin:#{app_data['app_env']['PATH']}"
  elsif node['opsline-rails-app']['ruby']['provider'] == 'rbenv'
    app_data['app_env']['PATH'] = "/home/#{node['opsline-rails-app']['owner']}/.rbenv/shims:/home/#{node['opsline-rails-app']['owner']}/.rbenv/bin:#{app_data['app_env']['PATH']}"
  end

  # merge inherited environment
  unless inherited_data.nil?
    if inherited_data.has_key?('environment')
      app_data['app_env'].merge!(get_env_value(inherited_data['environment']))
    end
    if inherited_data.has_key?('version')
      app_data['version'] = inherited_data['version']
    end
    if inherited_data.has_key?('type')
      app_data['type'] = inherited_data['type']
    end
    if inherited_data.has_key?('deploy_to')
      app_data['deploy_to'] = inherited_data['deploy_to']
    end
    if inherited_data.has_key?('package_type')
      app_data['package_type'] = inherited_data['package_type']
    end
    if inherited_data.has_key?('artifact_name')
      app_data['artifact_name'] = inherited_data['artifact_name']
    end
    if inherited_data.has_key?('artifact_location')
      app_data['artifact_location'] = inherited_data['artifact_location']
    end
    if inherited_data.has_key?('jenkins_job_name')
      app_data['jenkins_job_name'] = inherited_data['jenkins_job_name']
    end
  end

  # skip non-rails apps
  next unless app_data['type'] == 'rails'

  # set default parameters if not provided
  unless app_data.has_key?('deploy_to')
    app_data['deploy_to'] = "#{node['opsline-rails-app']['apps_root']}/#{app_name}"
  end
  unless app_data.has_key?('package_type')
    app_data['package_type'] = 'tar.gz'
  end
  unless app_data.has_key?('artifact_name')
    app_data['artifact_name'] = app_name
  end
  unless app_data.has_key?('artifact_location')
    app_data['artifact_location'] = "s3://s3.amazonaws.com/#{node['opsline-rails-app']['s3_bucket']}/#{app_data['artifact_name']}"
  end
  app_data['version'] = get_env_value(app_data['version'])
  app_data['artifact'] = "#{app_data['artifact_name']}-#{app_data['version']}.#{app_data['package_type']}"
  if app_data.has_key?('jenkins_job_name')
    app_data['artifact'] = "jobs/#{app_data['jenkins_job_name']}/#{app_data['version']}/#{app_data['artifact']}"
  end
  unless app_data.has_key?('container')
    app_data['container'] = 'unicorn'
  end
  if app_data.has_key?('container_parameters')
    app_data['container_parameters'] = get_env_value(app_data['container_parameters'])
  else
    app_data['container_parameters'] = {}
  end

  # merge environment from data bag
  if app_data.has_key?('environment')
    app_data['app_env'].merge!(get_env_value(app_data['environment']))
  end

  # create app directory
  directory app_data['deploy_to'] do
    action :create
    owner node['opsline-rails-app']['owner']
    group node['opsline-rails-app']['owner']
    mode 0755
  end

  # install pre-requisite packages
  if app_data.has_key?('packages')
    app_data['packages'].each do |pkg_name|
      package pkg_name do
        action :install
      end
    end
  end

  # configure logrotate
  unless app_data.has_key?('logrotate')
    logrotate = true
  else
    logrotate = to_bool(app_data['logrotate'])
  end
  if logrotate
    unless app_data.has_key?('logrotate_days')
      days = 7
    else
      days = app_data['logrotate_days'].to_i
    end
    logrotate_app app_name do
      cookbook 'logrotate'
      path "#{app_data['deploy_to']}/shared/log/*.log"
      options ['copytruncate', 'missingok', 'compress', 'notifempty', 'delaycompress']
      frequency 'daily'
      rotate days
    end
  end

  app_data['shared_directories'] = %w{ bundle config log pids sockets system }
  app_data['symlinks'] = {'log' => 'log'}
  # additional shared_subdirectories
  if app_data.has_key?('shared_subdirectories')
    app_data['shared_subdirectories'].each do |dir|
      app_data['shared_directories'] << dir
      app_data['symlinks'][dir] = dir
    end
  end

  services_to_restart = []
  pids_to_signal = []
  restart_after_deploy = false

  # deploy artifact
  artifact_deploy app_name do
    version app_data['version']
    artifact_location "#{app_data['artifact_location']}/#{app_data['artifact']}"
    deploy_to app_data['deploy_to']
    owner node['opsline-rails-app']['owner']
    group node['opsline-rails-app']['owner']
    environment app_data['app_env']
    shared_directories app_data['shared_directories']
    symlinks(app_data['symlinks'])
    action :deploy
    keep 2
    force false

    # remove log directory before linking
    before_symlink Proc.new {
      directory "#{app_data['deploy_to']}/releases/#{app_data['version']}/log" do
        action :delete
        recursive true
      end
    }

    # configure proc
    configure Proc.new {

      # UNICORN
      if app_data['container'] == 'unicorn'
        unless app_data['container_parameters'].has_key?('timeout')
          app_data['container_parameters']['timeout'] = node['opsline-rails-app']['unicorn']['timeout']
        end
        unless app_data['container_parameters'].has_key?('worker_processes')
          app_data['container_parameters']['worker_processes'] = node['opsline-rails-app']['unicorn']['worker_processes']
        end
        unless app_data['container_parameters'].has_key?('frontend')
          app_data['container_parameters']['frontend'] = 'nginx'
        end
        unless app_data['container_parameters'].has_key?('frontend_port')
          app_data['container_parameters']['frontend_port'] = '8080'
        end
        unicorn_config = "#{app_data['deploy_to']}/shared/config/unicorn.rb"

        env_dir app_data do
          deploy_to app_data['deploy_to']
          variables app_data['app_env']
          owner node['opsline-rails-app']['owner']
          group node['opsline-rails-app']['owner']
          notifies [:restart, "service[#{app_name}]"]
        end

        template unicorn_config do
          source 'unicorn.rb.erb'
          cookbook 'opsline-rails-app'
          owner node['opsline-rails-app']['owner']
          group node['opsline-rails-app']['owner']
          mode 0644
          action :create
          notifies :restart, "service[#{app_name}]"
          variables({
            :app_name => app_name,
            :user => node['opsline-rails-app']['owner'],
            :deploy_to => app_data['deploy_to'],
            :environment => app_data['app_env'],
            :parameters => app_data['container_parameters']
          })
        end

        template "/etc/init/#{app_name}.conf" do
          source 'upstart.unicorn.conf.erb'
          cookbook 'opsline-rails-app'
          owner 'root'
          group 'root'
          mode 0644
          action :create
          notifies :restart, "service[#{app_name}]"
          variables({
            :app_name => app_name,
            :user => node['opsline-rails-app']['owner'],
            :group => node['opsline-rails-app']['owner'],
            :deploy_to => app_data['deploy_to'],
            :exec_command => "unicorn -c #{unicorn_config}"
          })
        end
        service app_name do
          provider Chef::Provider::Service::Upstart
          action :nothing
        end
        pidfile = "#{app_data['deploy_to']}/shared/pids/unicorn.#{app_name}.pid"
        if File.exists?(pidfile)
          pids_to_signal << [File.read(pidfile).to_i, 'USR2']
        else
          services_to_restart << [app_name, Chef::Provider::Service::Upstart]
        end

        if app_data['container_parameters']['frontend'] == 'nginx'
          app_data['container_parameters']['upstream_sockets'] = [
            "#{app_data['deploy_to']}/shared/sockets/#{app_name}.socket"
          ]

          services_to_restart << ['nginx', Chef::Provider::Service::Init]

          nginx_app_config "nginx config for #{app_name}" do
            app_name app_name
            app_data app_data
            template 'nginx.unicorn.conf.erb'
          end
        end

      # PASSENGER
      elsif app_data['container'] == 'passenger'
        unless app_data['container_parameters'].has_key?('frontend')
          app_data['container_parameters']['frontend'] = 'nginx'
        end
        unless app_data['container_parameters'].has_key?('frontend_port')
          app_data['container_parameters']['frontend_port'] = '8080'
        end

        if app_data['container_parameters']['frontend'] == 'nginx'
          services_to_restart << ['nginx', Chef::Provider::Service::Init]

          nginx_app_config "nginx config for #{app_name}" do
            app_name app_name
            app_data app_data
            template 'nginx.passenger.conf.erb'
          end
        end

      # RACK
      elsif app_data['container'] == 'rack'
        env_dir app_data do
          deploy_to app_data['deploy_to']
          variables app_data['app_env']
          owner node['opsline-rails-app']['owner']
          group node['opsline-rails-app']['owner']
          notifies [:restart, "service[#{app_name}]"]
        end

        template "/etc/init/#{app_name}.conf" do
          source 'upstart.conf.erb'
          cookbook 'opsline-rails-app'
          owner 'root'
          group 'root'
          mode 0644
          action :create
          notifies :restart, "service[#{app_name}]"
          variables({
            :app_name => app_name,
            :user => node['opsline-rails-app']['owner'],
            :group => node['opsline-rails-app']['owner'],
            :deploy_to => app_data['deploy_to'],
            :exec_command => "rails server -P #{app_data['deploy_to']}/shared/pids/server.pid"
          })
        end
        service app_name do
          provider Chef::Provider::Service::Upstart
          action :nothing
        end
        services_to_restart << [app_name, Chef::Provider::Service::Upstart]

      # WORKER
      elsif app_data['container'] == 'worker'
        unless app_data['container_parameters'].has_key?('number_of_workers')
          app_data['container_parameters']['number_of_workers'] = 4
        end
        unless app_data['container_parameters'].has_key?('bundle_exec_command')
          app_data['container_parameters']['bundle_exec_command'] = ''
        end

        env_dir app_data do
          deploy_to app_data['deploy_to']
          variables app_data['app_env']
          owner node['opsline-rails-app']['owner']
          group node['opsline-rails-app']['owner']
          notifies [:restart, "service[#{app_name}]"]
        end

        template "/etc/init/#{app_name}-worker.conf" do
          source 'upstart.worker.conf.erb'
          cookbook 'opsline-rails-app'
          owner 'root'
          group 'root'
          mode 0644
          action :create
          notifies :restart, "service[#{app_name}]"
          variables({
            :app_name => app_name,
            :user => node['opsline-rails-app']['owner'],
            :group => node['opsline-rails-app']['owner'],
            :deploy_to => app_data['deploy_to'],
            :exec_command => app_data['container_parameters']['bundle_exec_command']
          })
        end
        template "/etc/init/#{app_name}.conf" do
          source 'upstart.workers.conf.erb'
          cookbook 'opsline-rails-app'
          owner 'root'
          group 'root'
          mode 0644
          action :create
          notifies :restart, "service[#{app_name}]"
          variables({
            :app_name => app_name,
            :user => node['opsline-rails-app']['owner'],
            :group => node['opsline-rails-app']['owner'],
            :deploy_to => app_data['deploy_to'],
            :number_of_workers => app_data['container_parameters']['number_of_workers']
          })
        end
        service app_name do
          provider Chef::Provider::Service::Upstart
          action :nothing
        end
        services_to_restart << [app_name, Chef::Provider::Service::Upstart]

      end
    }

    # restart proc
    restart Proc.new {
      restart_after_deploy = true
    }

    # after_deploy proc
    after_deploy Proc.new {
      if restart_after_deploy
        restart_services "restart #{app_name} services" do
          services_to_restart services_to_restart
          pids_to_signal pids_to_signal
        end
      end
    }

  end

end
